Add harmonic notch filter on accelerometer

guglie opened the pull request:
In my tests adding a filter on accelerometer improved the loiter performance on the altitude.
If there's a need to use less parameters we can use the same of the gyro, maybe just adding a flag to enable also on accelerometer.

andyp1per commented:
Nice! For what it's worth I think you could easily just use the current notch parameters. It's the same device so I think the frequency range with be the same between gyro and accel

auturgy commented:
Just to chip in: I don’t play in this area of ArduPilot, but I’m a geeky enough engineer to enjoy and appreciate when others do.
It’s cool to see you guys challenging, testing and evolving this stuff that is really quite deep down and complex.

Thanks!

James 

> On 6 Jun 2019, at 9:30 pm, Andy Piper <notifications@github.com> wrote:
> 
> Nice! For what it's worth I think you could easily just use the current notch parameters. It's the same device so I think the frequency range with be the same between gyro and accel
> 
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub, or mute the thread.

guglie commented:
@andyp1per yes, I think frequency is always the same, maybe attenuation can be different?
By the way the minimal parameter setup I can think of is to always apply notch on both gyro and accel (no parameters added or modified) or to make INS_NOTCH_ENABLE a bitmask (1 parameter sightly changed with backward compatibility)

@auturgy Thanks! I've done almost nothing, so I join you in thanking the other developers

magicrub commented:
>I've done almost nothing,

You're "almost nothing" is +56, −15 Lines of Code. That's definitely something :)

magicrub commented:
@guglie please be patient with this one, changes like this ripple into the entire system and every vehicle so it will likely be analyzed by several people before being merged. But that is a good thing, this change has the ability to have a large impact!

andyp1per commented:
@guglie I think you need to rebase on top of latest master - I see some of my changes to the notch configuration in here.

guglie commented:
@andyp1per it is based on latest master.
Your pull has been merged today! 😃

andyp1per commented:
Sorry my mistake 🙂 - you are right!

andyp1per commented:
Hey @guglie what hardware are you using, and are you using fast sampling? I think I have discovered a bug in the gyro configuration which means that accels are configured at full rate with no filtering and gyros are configured at full rate but with significant filtering. I'm wondering whether this is a reason that AltHold is affected. It might be worth trying to configure the accel like the gyro with sampling and see if that makes a difference.

guglie commented:
Hi @andyp1per, do you mean #11536 ?
I've tested both with an Invensense and a Bosch sensor, no I'm not using fast sampling.

lthall commented:
Tridge needs to check how expensive these changes are on processor time because they are being done at the IMU read rate. It would be nice to have this on all three accel axis if it is not too expensive!!

So we will need to wait for Tridge to do his thing.

guglie commented:
Thanks @lthall, surely @tridge review is needed also on parameter side.
Now this filter has it's own parameters (`NOTCA_` prefix), but what do you think about using same frequency, bandwidth and attenuation of gyro notch?

By the way I've been flying on a pixracer for almost 2 weeks with this accel notch on 3 axis, with loop rates above 1000 without problems.

lthall commented:
> what do you think about using same frequency, bandwidth and attenuation of gyro notch?

This is probably a good thing to do to start off with as it keeps things as simple as possible and minimises future parameter changes and conversions.

Ideally I would like to set up PID loops to have a section that runs at the gyro reed rate so we can define the notches per consumer rather than per source. That way they can be customized for each axis.

Assuming that Tridge gives the thumbs up I think using the same parameter is the best way to do this for now.

guglie commented:
@lthall I've set it up to use the same parameters. Now INS_NOTCH_ENABLE is a bitmask with the first bit enabling notch on gyro and the second enabling on accel. So it's backward compatible as a value of 1 enables it only on gyro, while setting INS_NOTCH_ENABLE = 3 enables both on gyro and accel using same parameters for frequency, bandwidth and attenuation.

Let me know if it's OK!

roque-canales commented:
Hello could you please confirm that this feature is activated on 3.6.10? 
Now can i use ins_notch_enable to 3 for activate acc filtering?

khancyr commented:
@roque-canales hello, this isn't merged yet. It doubt it comes before 3.7 serie

amilcarlucas commented:
This now has conflicts. Can you rebase it again?

guglie commented:
I've updated this PR with the (dynamic) harmonic notch filter on accelerometer.

As suggested by @lthall for the static notch, INS_HNTCH_ENABLE becomes a bitmask (backward compatible), setting it to 3 enables harmonic notch both on gyro and accel.

I've successfully tested this on Copter-4.0.2, but on master I get `undefined reference to enable_if/is_equal` error:  

    undefined reference to std::enable_if<std::is_floating_point<std::common_type<float, unsigned short>::type>::value, bool>::type is_equal<float, unsigned short>(float, unsigned short)

Any hint on this?
@andyp1per does it look good to you?

---
_As already discussed, the "old" static notch filter on accelerometer is also added by this PR:
INS_NOTCH_ENABLE becomes a bitmask (backward compatible), setting it to 3 enables static notch both on gyro and accel.
The other parameters for this notch are the same of the gyro._

andyp1per commented on libraries/AP_InertialSensor/AP_InertialSensor.cpp:
```diff
@@ -732,6 +732,14 @@ AP_InertialSensor::init(uint16_t sample_rate)
     }
+
+    for (uint8_t i=0; i<get_accel_count(); i++) {
+    	_accel_harmonic_notch_filter[i].allocate_filters(_harmonic_notch_filter.harmonics(),
```
I think by default we should allocate a single notch for accels and probably ignore this config option

guglie replied:
Ok, I agree it will free up some cpu, but how can we choose a single frequency?
(looking forward to your #14470)

andyp1per commented on libraries/AP_InertialSensor/AP_InertialSensor.h:
```diff
@@ -448,13 +449,15 @@ class AP_InertialSensor : AP_AccelCal_Client
+    // optional notch filter on gyro and accel
     NotchFilterParams _notch_filter;
     NotchFilterVector3f _gyro_notch_filter[INS_MAX_INSTANCES];
+    NotchFilterVector3f _accel_notch_filter[INS_MAX_INSTANCES];
```
Is this worth it? I would simply only provide the harmonic notch for accels

guglie replied:
This PR started with the static notch on accel, I would keep this one, is just another option for the user, the code is simple and tested and has a backward compatible param.
But let me know if there's a particular reason to remove it

andyp1per commented on libraries/AP_InertialSensor/AP_InertialSensor_Backend.cpp:
```diff
@@ -388,9 +388,26 @@ void AP_InertialSensor_Backend::_notify_new_accel_raw_sample(uint8_t instance,
 
-        _imu._accel_filtered[instance] = _imu._accel_filter[instance].apply(accel);
-        if (_imu._accel_filtered[instance].is_nan() || _imu._accel_filtered[instance].is_inf()) {
+        // apply the low pass filter
```
You will need to change this logic to match https://github.com/ArduPilot/ardupilot/pull/14470

andyp1per commented on libraries/AP_InertialSensor/AP_InertialSensor_Backend.cpp:
```diff
@@ -531,26 +548,26 @@ void AP_InertialSensor_Backend::update_gyro(uint8_t instance)
-    if (!is_equal(_last_notch_center_freq_hz, _gyro_notch_center_freq_hz()) ||
-        !is_equal(_last_notch_bandwidth_hz, _gyro_notch_bandwidth_hz()) ||
-        !is_equal(_last_notch_attenuation_dB, _gyro_notch_attenuation_dB()) ||
+    if (!is_equal(_last_gyro_notch_center_freq_hz, _notch_center_freq_hz()) ||
```
I know @rmackay9 objects to functions starting with _ so if you are going to change the names probably remove the _.

guglie replied:
Sure

andyp1per commented on libraries/AP_InertialSensor/AP_InertialSensor_Backend.h:
```diff
@@ -237,26 +237,29 @@ class AP_InertialSensor_Backend
 
     // return the notch filter center in Hz for the sample rate
-    float _gyro_notch_center_freq_hz(void) const { return _imu._notch_filter.center_freq_hz(); }
+    uint16_t _notch_center_freq_hz(void) const { return _imu._notch_filter.center_freq_hz(); }
```
Remove the _ and below

andyp1per commented on libraries/AP_InertialSensor/AP_InertialSensor_Backend.cpp:
```diff
@@ -574,6 +591,30 @@ void AP_InertialSensor_Backend::update_accel(uint8_t instance)
+	if (!is_equal(_last_harmonic_notch_bandwidth_hz, harmonic_notch_bandwidth_hz()) ||
+		!is_equal(_last_harmonic_notch_attenuation_dB, harmonic_notch_attenuation_dB()) ||
+		sensors_converging()) {
+		_imu._accel_harmonic_notch_filter[instance].init(_accel_raw_sample_rate(instance), harmonic_notch_center_freq_hz(), harmonic_notch_bandwidth_hz(), harmonic_notch_attenuation_dB());
```
These settings are synchronized - I would just move the update into the gyro update and be done with this check

andyp1per commented on libraries/AP_InertialSensor/AP_InertialSensor_Backend.h:
```diff
@@ -267,9 +270,14 @@ class AP_InertialSensor_Backend
-    float _last_notch_bandwidth_hz;
-    float _last_notch_attenuation_dB;
+    //gyro notch
+    float _last_gyro_notch_center_freq_hz;
```
Move update into gyro and remove the static accel notch and this can go away

andyp1per commented on libraries/Filter/NotchFilter.cpp:
```diff
@@ -100,8 +100,8 @@ const AP_Param::GroupInfo NotchFilterParams::var_info[] = {
-    // @Description: Enable notch filter
-    // @Values: 0:Disabled,1:Enabled
+    // @Description: Bitmask to enable notch filter on gyroscope and accelerometer.
+    // @Values: 0:Disabled,1:Enabled on gyroscope,2:Enabled on accelerometer,3:Enabled on gyroscope and accelerometer
```
Remove

amilcarlucas commented:
ping ....

khancyr commented:
@andyp1per @guglie do we still need this ?

davidbuzz commented:
lookinf for a comment from any of these people that this PR is still relevant...   @andyp1per , @tridge , @rmackay9

andyp1per commented:
Yeah I think it's still relevant but @guglie seems to have disappeared

peterbarker commented:
OK, massively conflicting, no response from author on the subject.

Closing this - but since Andy thinks this *might* be worthwhile, maybe open a fresh PR if this is still desired.
